### ISSUE TITLE
Add overload ImmutableArray.CreateRange(IEnumerable<T> items, Func<T,U> selector)

### ISSUE THREAD

**Issue by etishor** [dotnet/corefx#530]
## Scope
Add overload to the ImmutableArray.CreateRange method that would allow for reduced allocations when creating an ImmutableArray from an IEnumerable and a Func selector. 

## Rationale
This overload would allow for optimized array creation when the count of the items in the collection can be determined by the [TryGetCount()](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.cs#L26) method. 

Similar to the [overloads to ImmutableArray.CreateRange](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray.cs#L237) that take an ImmutableArray and a Func selector, another overload can be added to the CreateRange method:

```csharp
ImmutableArray<TResult> CreateRange<TSource, TResult>(IEnumerable<TSource> items, Func<TSource,TResult> selector)
```
The optimizations that can be performed are similar to the ones applied in the [overload that creates an ImmutableArray from an IEnumerable](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray.cs#L97)

## Usage

Some cases I've encountered, where it would be good if the extra allocations caused by calling LINQ ToArray() could be avoided:

```csharp
private List<string> data; // existing mutable data
public ImmutableArray<string> ReadTransformed()
{
    var transformed = data.Select(d => Transform(d)); //apply some transformation
    return ImmutableArray.CreateRange(transformed); // return as immutable
}
```
With the new overload the extra allocations could be avoided and this could be written like this:
```csharp
public ImmutableArray<string> ReadTransformed()
{
    return ImmutableArray.CreateRange(data, d => Transform(d));
}
```

## Implementation 

```csharp
public static ImmutableArray<TResult> CreateRange<TSource, TResult>(IEnumerable<TSource> items, Func<TSource, TResult> selector)
{
    Requires.NotNull(items, "items");

    int count;
    if (items.TryGetCount(out count))
    {
        var array = new TResult[count];
        int i = 0;
        foreach (var item in items)
        {
            array[i++] = selector(item);
        }

        return new ImmutableArray<TResult>(array);
    }
    else
    {
        return new ImmutableArray<TResult>(items.Select(selector).ToArray());
    }
}
```

## Other notes
I'm not sure if the use-case is not too specific to be in the public API. At the same time it is consistent with the existing overloads that create from ImmutableArray and feels like a good fit. To be even more consistent other overloads could be added that also take a start index and a length.
 
It might be possible to achieve the same result with the newly added ExtractToImmutable() method, by creating the builder with the known size and adding elements one by one and then extracting to immutable but that would require more code and would be error prone.

**Comment by AArnott** [comment]
Using the Builder might avoid the allocation, but creating the Builder itself is an allocation hit.
Also if you're trying to avoid a single allocation, the `Func` parameter alone might not cut it--we would probably want to add an overload that takes an argument to pass to the function as well.
But I'm not seeing the significant win here to justify 1-2 more method overloads. In the few cases where it would be useful, a shared/reused Builder instance would do the job as you say. And if initializing an array with a Builder is error prone, we have problems anyway since that's its main purpose.

**Comment by etishor** [comment]
I understand and agree that adding complexity to the API to save a few allocations is not necessarily desired. The ImmutableArray gives the user a very good balance between a safe and sane API while maintaining the performance of native arrays and I think a relevant part of its usage will be in high performance code. 

I've tried writing the code using a builder and this is what I ended up with:

```csharp
var builder = ImmutableArray.CreateBuilder<TResult>(count);
foreach (var item in items)
{
    builder.Add(selector(item));
}
return builder.MoveToImmutable();
```

I now realize that this is quite straight forward, and if the builder is cached it would have the same performance as the proposed overload. I'm also realizing that this is probably less error prone that having many overloads to a method. 

Any idea when the MoveToImmutable() will be in the published .nuget packages?

Thanks for the input,
Iulian
